src/matrixvariate.jl

```diff
@@ -124,45 +124,56 @@ end
 
 #### Sampling
 function Distributions._rand!(rng::AbstractRNG, d::TuringWishart, A::AbstractMatrix)
-    _wishart_genA!(rng, Distributions.dim(d), d.df, A)
+    Distributions._wishart_genA!(rng, Distributions.dim(d), d.df, A)
     unwhiten!(d.chol, A)
     A .= A * A'
 end
 
-function _wishart_genA!(rng::AbstractRNG, p::Int, df::Real, A::AbstractMatrix)
-    # Generate the matrix A in the Bartlett decomposition
-    #
-    #   A is a lower triangular matrix, with
-    #
-    #       A(i, j) ~ sqrt of Chisq(df - i + 1) when i == j
-    #               ~ Normal()                  when i > j
-    #
-    A .= zero(eltype(A))
-    for i = 1:p
-        @inbounds A[i,i] = rand(rng, Chi(df - i + 1.0))
-    end
-    for j in 1:p-1, i in j+1:p
-        @inbounds A[i,j] = randn(rng)
-    end
-end
-
 function unwhiten!(C::Cholesky, x::StridedVecOrMat)
     cf = C.U
     lmul!(transpose(cf), x)
 end
 
+## Custom adjoint since Zygote can't differentiate through `@warn`
+
+ZygoteRules.@adjoint function Wishart(df::T, S::AbstractPDMat{T}, warn::Bool = true) where T
+    return ZygoteRules.pullback(_Wishart, df, S, warn)
```

Comment:
Let's use a do block and define the function inside instead of outside.

Reply:
The reason why I didn't use anonymous functions is that it makes debugging more difficult and is therefore advised against in the ChainRules docs.

Reply:
I don't agree that it makes debugging more difficult. In this case, you would still get a useful stacktrace. It's just better to avoid additional function names if it is used in only one place.

Reply:
Hmm I'm not sure, the definition of `_Wishart` is quite long, so I guess it would be easier to keep it separate. Moreover, IMO it is easier to spot that it's just a copy of `Distributions.Wishart` with `@warn` replaced with `_warn`.

Reply:
@johnczito Could we just remove the warning in Distributions, as suggested in https://github.com/JuliaStats/Distributions.jl/pull/1097#discussion_r436377048?

Reply:
I just wanted to update the code and switch to a do block because it doesn't really matter to me in which way it's implemented, but I noticed that type annotations are not possible in do blocks (at least it failed when I tried a simple example with `map`). That means we would have to deviate from the original constructor in Distributions more drastically by removing the type annotations and retrieving the types inside of the function body instead - should we do that?

Reply:
I think we can just use the type parameters from the `@adjoint` function header.